pkg/pillar/zedbox/zedbox.go

```diff
@@ -33,46 +41,237 @@ import (
+		}
+	}
+}
+
+//startSubscriber Creates a socket for he agent and start listening
+func startSubscriber(agent string, topic interface{}, subSkt chan string) {
+	log.Infof("startSubscriber(%s)", agent)
+	sockName := getSocketName(agent, topic)
+	dir := path.Dir(sockName)
+	if _, err := os.Stat(dir); err != nil {
+		log.Infof("startSubscriber(%s): Create %s\n", agent, dir)
+		if err := os.MkdirAll(dir, 0700); err != nil {
+			log.Fatalf("startSubscriber(%s): Exception while creating %s. %s",
+				agent, dir, err)
+		}
+	}
+	if _, err := os.Stat(sockName); err == nil {
+		// This could either be a left-over in the filesystem
+		// or some other process (or ourselves) using the same
+		// name to publish. Try connect to see if it is the latter.
+		_, err := net.Dial("unix", sockName)
+		if err == nil {
+			log.Fatalf("connectAndRead(%s): Can not publish %s since its already used",
+				agent, sockName)
+		}
+		if err := os.Remove(sockName); err != nil {
+			log.Fatalf("connectAndRead(%s): Exception while removing pre-existing sock %s. %s",
+				agent, sockName, err)
+		}
+	}
+	listener, err := net.Listen("unix", sockName)
+	if err != nil {
+		log.Fatalf("connectAndRead(%s): Exception while listening at sock %s. %s",
+			agent, sockName, err)
+	}
+	defer listener.Close()
+	for {
+		c, err := listener.Accept()
+		if err != nil {
+			log.Errorf("connectAndRead(%s) failed %s\n", sockName, err)
+			continue
+		}
+		go serveConnection(c, subSkt)
+	}
+}
+
+//publish publishes data to an already opened socket.
+func publish(agent string, data interface{}) error {
+	log.Infof("publish(%s)", agent)
+	sockName := getSocketName(agent, data)
+
+	if _, err := os.Stat(sockName); err != nil {
+		err := fmt.Errorf("publish(%s): exception while check socket. %s", sockName, err.Error())
+		log.Errorf(err.Error())
+		return err
+	}
+
+	byteData, err := json.Marshal(data)
+	if err != nil {
+		err := fmt.Errorf("publish(%s): exception while marshalling data. %s",
+			sockName, err.Error())
+		log.Errorf(err.Error())
+		return err
+	}
+
+	conn, err := net.Dial("unix", sockName)
+	if err != nil {
+		err := fmt.Errorf("publish(%s): exception while dialing socket. %s",
+			sockName, err.Error())
+		log.Errorf(err.Error())
+		return err
+	}
+
+	if _, err := conn.Write(byteData); err != nil {
+		err := fmt.Errorf("publish(%s): exception while writing data to the socket. %s",
+			sockName, err.Error())
+		log.Errorf(err.Error())
+		return err
+	}
+	return nil
+}
+
+func serveConnection(conn net.Conn, retChan chan string) {
+
+	for {
+		buf := make([]byte, 2048)
+		count, err := conn.Read(buf)
+		if err != nil {
+			log.Errorf("serveConnection: Error on read: %s", err)
+			break
+		}
+		retChan <- string(buf[:count])
+
+	}
+}
+
+func getSocketName(agent string, topic interface{}) string {
+	return path.Join(socketPath, agent, fmt.Sprintf("%s.sock", pubsub.TypeToName(topic)))
+}
+
+//startAgentAndDone start the given agent. Touches a <agentName>.done file once the agent returns.
```

Comment:
Question; if the agent aka zedbox hits a log.Fatal or panic I assume the serviceEntrypoint aka Run() will not return. Is that correct?

Reply:
Yeah.